src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -1647,28 +1809,75 @@ - (void)_handleReportEnd
+        if ([testDelegate respondsToSelector:@selector(unitTestReportEndForDevice:)]) {
+            [testDelegate unitTestReportEndForDevice:self];
+        }
+    }];
+#endif
+}
+
+- (BOOL)_interestedPaths:(NSArray * _Nullable)interestedPaths includesPath:(MTRAttributePath *)attributePath
+{
+    for (id interestedPath in interestedPaths) {
+        if ([interestedPath isKindOfClass:[MTRClusterPath class]]) {
+            MTRClusterPath * interestedClusterPath = interestedPath;
+            if (interestedClusterPath.cluster.unsignedLongValue == attributePath.cluster.unsignedLongValue) {
+                return YES;
             }
-        });
+        } else if ([interestedPath isKindOfClass:[MTRAttributePath class]]) {
+            MTRAttributePath * interestedAttributePath = interestedPath;
+            if ((interestedAttributePath.cluster.unsignedLongValue == attributePath.cluster.unsignedLongValue) && (interestedAttributePath.attribute.unsignedLongValue == attributePath.attribute.unsignedLongValue)) {
+                return YES;
+            }
+        }
     }
-#endif
+
+    return NO;
+}
+
+// Returns filtered set of attributes using an interestedPaths array.
+// Returns nil if no attribute reports has a path that matches the paths in the interestedPaths array.
+- (NSArray<NSDictionary<NSString *, id> *> *)_filteredAttributes:(NSArray<NSDictionary<NSString *, id> *> *)attributes forInterestedPaths:(NSArray * _Nullable)interestedPaths
+{
+    if (!interestedPaths.count) {
+        return attributes;
+    }
+
+    NSMutableArray * filteredAttributes = nil;
+    for (NSDictionary<NSString *, id> * responseValue in attributes) {
+        MTRAttributePath * attributePath = responseValue[MTRAttributePathKey];
+        if ([self _interestedPaths:interestedPaths includesPath:attributePath]) {
+            if (!filteredAttributes) {
+                filteredAttributes = [NSMutableArray array];
+            }
+            [filteredAttributes addObject:responseValue];
+        }
+    }
+
+    if (filteredAttributes.count && (filteredAttributes.count != attributes.count)) {
+        MTR_LOG("%@ filtered attribute report %lu => %lu", self, static_cast<unsigned long>(attributes.count), static_cast<unsigned long>(filteredAttributes.count));
+    }
+
+    return filteredAttributes;
 }
 
 // assume lock is held
 - (void)_reportAttributes:(NSArray<NSDictionary<NSString *, id> *> *)attributes
 {
     os_unfair_lock_assert_owner(&self->_lock);
     if (attributes.count) {
-        id<MTRDeviceDelegate> delegate = _weakDelegate.strongObject;
-        if (delegate) {
-            dispatch_async(_delegateQueue, ^{
-                [delegate device:self receivedAttributeReport:attributes];
-            });
-        }
+        [self _iterateDelegatesWithBlock:^(MTRDeviceDelegateInfo * delegateInfo) {
+            // make an autorelease pool so that temporary filtered attributes reports don't bloat memory
+            @autoreleasepool {
```

Comment:
It would probably be good to just have an `@autoreleasepool` in `_iterateDelegatesWithBlock` so it applies to all these iterations... auto release pool blocks are cheap.